Add readme for tests (#554) (#555)
knative/client: test/README.md

# Test

This directory contains tests and testing docs for `Kn client`:

- [Unit tests](#running-unit-tests) currently reside in the codebase alongside
  the code they test
- [End-to-end tests](#running-end-to-end-tests) :
  - end-to-end tests in [`/test/e2e`](./e2e)

## Running unit tests

To run all unit tests:

```bash
$ hack/build.sh -f -t
```

_By default `go test` will not run [the e2e tests](#running-end-to-end-tests-locally),
which need [`-tags=e2e`](#running-end-to-end-tests) to be enabled._

## Running end to end tests locally

To run [the e2e tests](./e2e) , you need to have a 

1. [running knative environment.](./../DEVELOPMENT.md#create-a-cluster)
2. kn binary in the $path. 

Before running the e2e tests please make sure you dont have any namespaces with the name starting with `kne2etests`

```bash
$ ./e2e-tests-local.sh
```

> 

Reviewer: please add running e2e tests selectively using tags as mentioned in the PR description at https://github.com/knative/client/pull/547#issue-351756917

# Test

This directory contains e2e tests and testing docs:

- Unit tests are in the code base alongside the code they test
- e2e tests are in [`test/e2e/`](./e2e)

## Running unit tests

To run all unit tests:

```bash
$ hack/build.sh -t
```

_By default `go test` will not run [the e2e tests](#running-e2e-tests-locally),
which need [`-tags=e2e`](#running-end-to-end-tests) to be enabled._

## Running e2e tests locally

To run [the e2e tests](./e2e) , you need to have a 

1. [Running knative environment.](./../DEVELOPMENT.md#create-a-cluster)
2. `kn` binary in the $PATH.

> Before running the e2e tests please make sure you dont have any namespaces with the name starting with `kne2etests`

Reviewer 2: Might want to mention that their `KUBECONFIG` env variable needs to be set or that `~/.kube/config` file is pointing to the cluster's configuration.

Author: Added

- Unit tests are in the code base alongside the code they test
- e2e tests are in [`test/e2e/`](./e2e)

## Running unit tests

To run all unit tests:

```bash
$ hack/build.sh -t
```

_By default `go test` will not run [the e2e tests](#running-e2e-tests-locally),
which need [`-tags=e2e`](#running-end-to-end-tests) to be enabled._

## Running e2e tests locally

To run [the e2e tests](./e2e) , you need to have a 

1. [Running knative environment.](./../DEVELOPMENT.md#create-a-cluster)
2. `kn` binary in the $PATH.
3. Please Make sure that you are able to connect to the cluster by following the [guide here](./../docs#connecting-to-your-cluster)

Before running the e2e tests please make sure you dont have any namespaces with the name starting with `kne2etests`

Run all e2e tests:

```bash
$ test/e2e-tests-local.sh
```

### Running e2e tests selectively

To run only serving specific e2e tests locally, use

```bash
E2E_TAGS="serving" test/e2e-tests-local.sh
```

To run only eventing specific e2e tests locally, use

```bash
E2E_TAGS="eventing" test/e2e-tests-local.sh
```

### Running a single test case

To run one e2e test case, e.g. TestBasicWorkflow

```bash

> go test -v -tags=e2e -count=1 ./e2e -run ^TestBasicWorkflow$

Reviewer: We can do this as suggested `test/e2e-tests-local.sh -run ^TestBasicWorkflow$`

Author: @navidshaikh thank you.
Instead of 'go test' use the script ?

Reviewer: Yes, see https://github.com/knative/client/pull/555#discussion_r358082251

# Test

> This directory contains tests and testing docs for `Kn client`:

Reviewer: ```suggestion
This directory contains e2e tests and testing docs:
```

> - [Unit tests](#running-unit-tests) currently reside in the codebase alongside

Reviewer: ```suggestion
- Unit tests are in the code base alongside the code they test
```

  the code they test

> - [End-to-end tests](#running-end-to-end-tests) :

Reviewer: ```suggestion
- e2e tests are in [`test/e2e/`](./e2e)
```

  - end-to-end tests in [`/test/e2e`](./e2e)

## Running unit tests

To run all unit tests:

```bash

> $ hack/build.sh -f -t

Reviewer: ```suggestion
$ hack/build.sh -t
```

```

_By default `go test` will not run [the e2e tests](#running-end-to-end-tests-locally),
which need [`-tags=e2e`](#running-end-to-end-tests) to be enabled._

> ## Running all end to end tests locally

Reviewer: ```suggestion
## Running e2e tests locally
```

To run [the e2e tests](./e2e) , you need to have a 

1. [running knative environment.](./../DEVELOPMENT.md#create-a-cluster)

> 2. kn binary in the $path. 

Reviewer: ```suggestion
2. `kn` binary in the $PATH.
```

Before running the e2e tests please make sure you dont have any namespaces with the name starting with `kne2etests`

> 

Reviewer: ```suggestion
Run all e2e tests:
```

```bash

> $ ./e2e-tests-local.sh

Reviewer: ```suggestion
$ test/e2e-tests-local.sh
```

```

> ### Running end to end tests selectively

Reviewer: ```suggestion
### Running e2e tests selectively
```

To run only serving specific e2e tests locally, use

```

> E2E_TAGS="serving" ./e2e-tests-local.sh

Reviewer: ```suggestion
E2E_TAGS="serving" test/e2e-tests-local.sh
```

```

To run only eventing specific e2e tests locally, use

```

> E2E_TAGS="eventing" ./e2e-tests-local.sh

Reviewer: ```suggestion
E2E_TAGS="eventing" test/e2e-tests-local.sh
```

```

### Running a single test case

To run one e2e test case, e.g. TestAutoscaleUpDownUp, use
[the `-run` flag with `go test`](https://golang.org/cmd/go/#hdr-Testing_flags):

```bash
go test -v -tags=e2e -count=1 ./e2e -run ^TestBasicWorkflow$

> ```

Reviewer: lets swap this with following text:

To run specific e2e test case, e.g. TestBasicWorkflow
```
test/e2e-tests-local.sh -run ^TestBasicWorkflow$
```

Author: Thanks for reviewing. I have made the changes.

### Running tests in short mode

Running tests in short mode excludes some large-scale E2E tests and saves
time/resources required for running the test suite. To run the tests in short
mode, use
[the `-short` flag with `go test`](https://golang.org/cmd/go/#hdr-Testing_flags)

```bash

> go test -v -tags=e2e -count=1 -short ./e2e

Reviewer: ```suggestion
test/e2e-tests-local.sh -short
```

Reviewer: Good point!
Although we don't have e2e tests responding to `-short` flag, but I'd keep this section. we can plan to skip some long running e2e tests for short mode.

```

## Presubmit tests

> Presubmit tests and subsequents tests require a --gcp-project and cannot be run locally.

Reviewer: Please remove this section, since those checks are in `./hack/build.sh` (see `./hack/build.sh -h`)

Author: removed the section
